### ISSUE TITLE
[Bug][broker] Error reading entries due to LastConfirmedEntry is null until broker restart

### ISSUE THREAD

**Issue by PatrykWitkowski** [apache/pulsar#23654]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Read release policy

- [X] I understand that unsupported versions don't get bug fixes. I will attempt to reproduce the issue on a supported version of Pulsar client and Pulsar broker.

### Version

OS: docker image apachepulsar/pulsar:3.3.2
Java: Amazon Corretto 21
Pulsar: 3.3.2 (client + server)

### Minimal reproduce step

Happens randomly during redeployment of a pulsar cluster - restarting all pulsar cluster components (brokers, bookies, zookeepers).

### What did you expect to see?

No mentioned error appear on brokers after restarting them.

### What did you see instead?

After brokers restart we start to see a lot of errors on brokers:
`[persistent://datalake/ingress/...-partition-... / sub] Error reading entries at 3904681:27 : LastConfirmedEntry is null when reading ledger 3904681, Read Type Normal - Retrying to read in 56.202 seconds"`
Retrying to read keeps happen until the next brokers restart.

At the same time we see growing backlog on that topic as the consumer can't read
![image](https://github.com/user-attachments/assets/1315d718-5dd5-4efc-84b8-bbc1a5a60bd4)

It happens on multipartition topic with the consumer from this class: org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers

When we detect it and restart all brokers once again then the backlog goes down (as you can see on the screenshot) and mentioned error goes away. So there's a workaround for that issue.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by lhotari** [comment]
This is kind of surprising logic:
https://github.com/apache/pulsar/blob/7e6fa554012b237308db1bebf3c625e103bf0db5/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L588-L599

Perhaps it's correct, but I'd assume that `lastConfirmedEntry` would be set earlier and not modified after that.
 
https://github.com/apache/pulsar/blob/7822dca1ffe45324d7af7ef830c617f6881a5431/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L430-L433

It looks like the iteration logic was added in #1550. Just wondering if it really makes sense in all cases.

**Comment by BewareMyPower** [comment]
Yes, https://github.com/apache/pulsar/pull/23147 adds a compare logic with the last confirmed entry so that it assumes the `lastConfirmedEntry` is not null. However, this case is impossible once after the `initializeBookKeeper` method is called because this field will be initialized with a non-null value and all subsequent modifications do not change it to null.

**Comment by BewareMyPower** [comment]
https://github.com/apache/pulsar/pull/23368 might fix this issue. I observed a similar issue before. From the heap dump I saw the managed ledger held by `RangeCacheImpl` is outdated and invalid.

**Comment by PatrykWitkowski** [comment]
@lhotari In our case that issue was taking 10 days (15/11/2024 - 25/11/2024) and it eventually recovered but **after brokers restart**

@BewareMyPower I see that fix is part of 4.0.0 so migrating to that version might solve it?

**Comment by BewareMyPower** [comment]
From that PR, you can see the `release/3.3.3` label, which means it's included in the 3.3.3 release. So you can upgrade the broker from 3.3.2 to 3.3.3
